cmd/soroban-cli/src/snapshot.rs

```diff
@@ -120,10 +127,95 @@ where
         protocol_version: ledger_info.protocol_version,
         sequence_number: ledger_info.sequence_number,
         timestamp: ledger_info.timestamp,
-        network_passphrase: ledger_info.network_passphrase,
+        network_passphrase: ledger_info.network_passphrase.clone(),
         base_reserve: ledger_info.base_reserve,
     };
     serde_json::to_writer(&file, &output)?;
 
     Ok(())
 }
+
+/// Returns a list of events from the on-disk event store, which stores events
+/// exactly as they'd be returned by an RPC server.
+pub fn read_events(path: &std::path::PathBuf) -> Result<Vec<rpc::Event>, Error> {
+    let reader = std::fs::OpenOptions::new().read(true).open(path)?;
+    Ok(serde_json::from_reader(reader)?)
+}
+
+// Reads the existing event file, appends the new events, and writes it all to
+// disk. Note that this almost certainly isn't safe to call in parallel.
+pub fn commit_events(
+    new_events: &[events::HostEvent],
+    ledger_info: &LedgerInfo,
+    output_file: &std::path::PathBuf,
+) -> Result<(), Error> {
+    // Create the directory tree if necessary, since these are unlikely to be
+    // the first events.
+    if let Some(dir) = output_file.parent() {
+        if !dir.exists() {
+            create_dir_all(dir)?;
+        }
+    }
+
+    let mut file = std::fs::OpenOptions::new()
+        .create(true)
+        .read(true)
+        .open(output_file)?;
+    let mut events: rpc::GetEventsResponse = serde_json::from_reader(&mut file)?;
+
+    for (i, event) in new_events.iter().enumerate() {
+        let contract_event = match event {
+            events::HostEvent::Contract(e) => e,
+            events::HostEvent::Debug(_e) => todo!(),
+        };
+
+        let topics = match &contract_event.body {
+            xdr::ContractEventBody::V0(e) => &e.topics,
+        }
+        .iter()
+        .map(xdr::WriteXdr::to_xdr_base64)
+        .collect::<Result<Vec<String>, _>>()?;
+
+        let id = toid::Toid::new(
+            ledger_info.sequence_number,
+            // we should technically inject the tx order here from the ledger
+            // info, but the sandbox does one tx/op per ledger anyway, so this
+            // is a safe assumption
+            1,
+            1,
+        );
+        let cereal_event = rpc::Event {
+            event_type: "contract".to_string(),
+            id: id.to_string(),
+            // stolen from
+            // https://github.com/stellar/soroban-tools/blob/main/cmd/soroban-rpc/internal/methods/get_events.go#L264
+            paging_token: format!("{}-{:010}", id.to_paging_token(), i + 1),
+            ledger: ledger_info.sequence_number.to_string(),
+            ledger_closed_at: ledger_info.timestamp.to_string(),
```

Comment:
*question*
Unless I'm misreading this, this is a `u64::to_string()`? Which means we're re-using `rpc::Event` here, but serializing stuff differently. In events from the soroban-rpc server this is an ISO8601 string, but here this is a stringified u64. It would be less confusing in future if we are consistent, so can we serialize this as an iso8601?

Edit: In the test fixtures file it is an iso8601, so.. what?

Reply:
Yeah I bamboozled myself by creating the fixture file manually :yum: Fixed in 39376cfa33e464afd823efb93fec32fc5727d270!